lib/msf/base/simple/job_state_tracker.rb

```diff
@@ -0,0 +1,80 @@
+require 'monitor'
+
+class JobStateTracker
+
+  include MonitorMixin
+
+  def initialize(result_ttl=nil)
+    self.results_size = 0
+    self.ready = Set.new
+    self.running = Set.new
+    # Can be expanded upon later to allow the option of a MemCacheStore being backed by redis for example
+    self.results = ActiveSupport::Cache::MemoryStore.new(expires_in: result_ttl || 5.minutes)
+  end
+
+  def waiting(id)
+    ready << id
+  end
+
+  def start(id)
+    running << id
+    ready.delete(id)
+  end
+
+  def completed(id, result, ttl=nil)
+    begin
+      # ttl of nil means it will take the default expiry time
+      self.results_size += 1 if results.write(id, {result: result}, ttl)
```

Comment:
Correct me if i'm wrong, but will the value of `results_size` be wrong when the memory store expires its values behind the scenes?

Reply:
you are 100% correct...I...don't have a way to solve that problem, not without reaching into places I shouldn't be in the cache